File: 22-generate-parentheses/22-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack, res = [], []
        
        def generate(openN, closedN):
            
            if openN == closedN == n:
                res.append("".join(stack))
            
            if openN < n:
                stack.append('(')
                generate(openN + 1, closedN)
                stack.pop()
            
            if closedN < openN:
                stack.append(')')
                generate(openN, closedN + 1)
                stack.pop()
                
        generate(0,0)
        return res

File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    vector<string> res;
    vector<char> cur;
    void dfs(int left, int right, int n) {
        if (left == right && right == n) {
            string s(cur.begin(), cur.end());
            res.push_back(s);
            return;
        }
        if (left < n) {
            cur.push_back('(');
            dfs(left + 1, right, n);
            cur.pop_back();
        }
        if (right < left) {
            cur.push_back(')');
            dfs(left, right + 1, n);
            cur.pop_back();
        }
    }
    vector<string> generateParenthesis(int n) {
        dfs(0, 0, n);
        return res;
    }
};